linera-views/src/views/bucket_queue_view.rs

```diff
@@ -0,0 +1,700 @@
+        let mut stored_data = match front {
+            Some(front) => {
+                vec![Some(front)]
+            }
+            None => {
+                vec![]
+            }
+        };
+        let stored_indices = from_bytes_option_or_default::<StoredIndices, _>(value2)?;
+        for _ in 1..stored_indices.len() {
+            stored_data.push(None);
+        }
+        let cursor = Cursor::new(&stored_indices);
+        Ok(Self {
+            context,
+            stored_data,
+            new_back_values: VecDeque::new(),
+            stored_indices,
+            cursor,
+            delete_storage_first: false,
+        })
+    }
+
+    async fn load(context: C) -> Result<Self, ViewError> {
+        let keys = Self::pre_load(&context)?;
+        let values = context.read_multi_values_bytes(keys).await?;
+        Self::post_load(context, &values)
+    }
+
+    fn rollback(&mut self) {
+        self.delete_storage_first = false;
+        self.cursor = Cursor::new(&self.stored_indices);
+        self.new_back_values.clear();
+    }
+
+    async fn has_pending_changes(&self) -> bool {
+        if self.delete_storage_first {
+            return true;
+        }
+        if !self.stored_indices.is_empty() {
+            let Some((i_block, position)) = self.cursor.position else {
+                return true;
+            };
+            if i_block != 0 || position != self.stored_indices.position {
+                return true;
+            }
+        }
+        !self.new_back_values.is_empty()
+    }
+
+    fn flush(&mut self, batch: &mut Batch) -> Result<bool, ViewError> {
+        let mut delete_view = false;
+        assert_eq!(self.stored_data.len(), self.stored_indices.len());
+        if self.delete_storage_first {
+            let key_prefix = self.context.base_key();
+            batch.delete_key_prefix(key_prefix);
+            delete_view = true;
+        }
+        if self.stored_count() == 0 {
+            let key_prefix = self.context.base_key();
+            batch.delete_key_prefix(key_prefix);
+            self.stored_indices = StoredIndices::default();
+            self.stored_data.clear();
+        } else if let Some((i_block, position)) = self.cursor.position {
+            for block in 0..i_block {
+                let index = self.stored_indices.indices[block].1;
+                let key = self.get_index_key(index)?;
+                batch.delete_key(key);
+            }
+            self.stored_indices.indices.drain(0..i_block);
+            self.stored_data.drain(0..i_block);
+            self.cursor = Cursor {
+                position: Some((0, position)),
+            };
+            // We need to ensure that the first index is in the front.
+            let first_index = self.stored_indices.indices[0].1;
+            if first_index != 0 {
+                let size = self.stored_indices.indices[0].0;
+                self.stored_indices.indices[0] = (size, 0);
```

Comment:
But why do we bother at all enforcing that the first index is always 0?

Why not instead keep the indices consecutive? Then we also wouldn't have to store each index, only the first one?

Reply:
Yes, we want to preload the data right away.
Yes, that code has been corrected away.